### ISSUE TITLE
Interface field resolving causes "Concrete type is not handled by instance resolvers on GraphQL union"

### ISSUE THREAD

**Issue by acelot** [graphql-rust/juniper#798]
_The bug initially was found while using [juniper-from-schema](https://github.com/davidpdrsn/juniper-from-schema) library. Additional information can be found in [this issue](https://github.com/davidpdrsn/juniper-from-schema/issues/130). Nevertheless, I will repeat the description of the error here:_

I have some field that returns an union type:
```graphql
type Query {
  testField: TestFieldResult!
}

union TestFieldResult =
  | Ok
  | SomeProblem
  | AnotherProblem

type Ok {
  dummy: Boolean!
}
```

`SomeProblem` and `AnotherProblem` are implementing one shared interface:
```graphql
type SomeProblem implements ProblemInterface {
  message: String!
}

type AnotherProblem implements ProblemInterface {
  message: String!
}

interface ProblemInterface {
  message: String!
}
```

Now I'm trying to perform the query and I just need to know that there is a problem, I am not interested in a specific one:
```graphql
query {
  testField {
    __typename
    ... on Ok {
      dummy
    }
    ... on ProblemInterface {
      message
    }
  }
}
```

And I getting an error:
```
thread 'actix-rt:worker:2' panicked at 'Concrete type ProblemInterface is not handled by instance resolvers on GraphQL union TestFieldResult'
```

However, if you specify a specific type of problem, then everything works.

Gist to reproduce an error: https://gist.github.com/acelot/0f54704452b9cff373d280ea3ee94d81

**Comment by tyranron** [comment]
@acelot this is not a bug. What you want is not allowed by [GraphQL spec](https://spec.graphql.org/June2018/#sec-Unions).

Given the syntax `...on Type` on GraphQL union means that you try to downcast the union into its variant (member). However, the spec is quite explicit about what types can represent GraphQL union variants:
> 2. The member types of a Union type must all be Object base types; Scalar, Interface and Union types must not be member types of a Union. Similarly, wrapping types must not be member types of a Union.

So you cannot specify GraphQL interface in `...on Type` syntax on the GraphQL union, nor you can make a GraphQL interface a variant of the union. GraphQL is not intended to work like that, sorry 🤷‍♂️

But such situation should definitely not result in a panic. Need to tune up validation rules.

**Comment by acelot** [comment]
it's a shame it doesn't work that way.  Didn't know it was written in the spec.  Strange, but other libraries support this feature like `gqlgen` and `HotChocolate`.

**Comment by tyranron** [comment]
@acelot examining it more deeply, seems that I was wrong. Spec has [only this about spreading interfaces on unions and vice versa](https://spec.graphql.org/June2018/#sec-Abstract-Spreads-in-Abstract-Scope):
> Union or interfaces fragments can be used within each other. As long as there exists at least one object type that exists in the intersection of the possible types of the scope and the spread, the spread is considered valid.
>
>So for example
>
> ```graphql
> fragment unionWithInterface on Pet {
>   ...dogOrHumanFragment
> }
> 
> fragment dogOrHumanFragment on DogOrHuman {
>   ... on Dog {
>     barkVolume
>   }
> }
> ```
> is consider valid because Dog implements interface Pet and is a member of DogOrHuman.

So, should figure out how to fix this...

**Comment by acelot** [comment]
Any news on this?

**Comment by tyranron** [comment]
@acelot nope, haven't yet time to dive into this.

**Comment by BeneSim** [comment]
Just wanted to add that the "learn" documentation on graphql.org also deals with this when discussing unions
https://graphql.org/learn/schema/#union-types
> Also, in this case, since Human and Droid share a common interface (Character), you can query their common fields in one place rather than having to repeat the same fields across multiple types:
```{
  search(text: "an") {
    __typename
    ... on Character {
      name
    }
    ... on Human {
      height
    }
    ... on Droid {
      primaryFunction
    }
    ... on Starship {
      name
      length
    }
  }
}```
Need to apply some workarounds in my schema due to this "bug", so count me in for the ones that would like to see this fixed :)

**Comment by acelot** [comment]
@BeneSim agree with you. For me, the absence of this feature is a show stopper, because all my API is focused on interfaces.

**Comment by tyranron** [comment]
@acelot @BeneSim fine... just started looking into this. Not yet any estimates can be given, but it's on my focus now.

**Comment by tyranron** [comment]
@acelot @BeneSim it should be OK now on `master`. The fix is easier than I thought it would be. I'll issue the new `juniper` release in few days.

**Comment by acelot** [comment]
@tyranron Can't wait for the release to check out the fix :)

**Comment by tyranron** [comment]
@acelot well, that's a bit unfortuned. The fix is scheduled for the next major 0.16 release, which will land in 2-3 weeks or so, but with nicer reworked macros and `tracing` support.

Sorry for that.
